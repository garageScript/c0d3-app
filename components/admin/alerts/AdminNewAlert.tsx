import { useMutation } from '@apollo/client'
import React, { useState, useEffect } from 'react'
import * as Sentry from '@sentry/browser'
import Alert from '../../../components/Alert'
import { FormCard } from '../../../components/FormCard'
import ADD_ALERT from '../../../graphql/queries/addAlert'
import { Alert as AlertType } from '../../../graphql'
import noop from '../../../helpers/noop'
import { alertValidation } from '../../../helpers/formValidation'
import { formChange } from '../../../helpers/formChange'
import {
  errorCheckAllFields,
  getPropertyArr,
  makeGraphqlVariable
} from '../../../helpers/admin/adminHelpers'

type NewAlertProps = {
  setAlerts: React.Dispatch<React.SetStateAction<AlertType[]>>
}

const newAlertAttributes = {
  text: '',
  type: [
    { title: 'info', as: 'button' },
    { title: 'urgent', as: 'button' }
  ],
  url: '',
  urlCaption: ''
}

export const NewAlert: React.FC<NewAlertProps> = ({ setAlerts }) => {
  const handleChange = async (value: string, propertyIndex: number) => {
    await formChange(
      value,
      propertyIndex,
      alertProperties,
      setAlertProperties,
      alertValidation
    )
  }

  const blankValuesForm = getPropertyArr(
    { ...newAlertAttributes },
    undefined,
    handleChange
  )

  const [createAlert, { loading, data }] = useMutation(ADD_ALERT)
  const [alertProperties, setAlertProperties] = useState(blankValuesForm)

  // when data is fully loaded after sending mutation request, update front-end alerts info
  useEffect(() => {
    if (!loading && data) {
      setAlerts(data.addAlert)
      setAlertProperties(blankValuesForm)
    }
  }, [data])

  // alter gets called when someone clicks button to create a lesson
  const alter = async () => {
    const newProperties = [...alertProperties]
    const valid = await errorCheckAllFields(newProperties, alertValidation)
    if (!valid) {
      setAlertProperties(newProperties)
      return
    }
    try {
      await createAlert(makeGraphqlVariable(newProperties))
    } catch (err) {
      Sentry.captureException(err)
    }
  }

  return (
    <div className="row mb-4">
      <div className="col-8 offset-2">
        <div className="mb-3">
          <Alert
            alert={{
              id: '-1',
              ...makeGraphqlVariable(alertProperties).variables
            }}
            onDismiss={noop}
          />
        </div>
        <div className="text-left">
          <FormCard
            values={alertProperties}
            onSubmit={{ title: 'Create New Alert', onClick: alter }}
            onChange={handleChange}
          />
        </div>
      </div>
    </div>
  )
}
