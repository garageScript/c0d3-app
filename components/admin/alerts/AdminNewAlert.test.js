import * as React from 'react'
import { MockedProvider } from '@apollo/client/testing'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { NewAlert } from './AdminNewAlert'
import ADD_ALERT from '../../../graphql/queries/addAlert'
import * as Sentry from '@sentry/browser'
jest.mock('@sentry/browser')
const addAlertMock = {
  request: {
    query: ADD_ALERT,
    variables: {
      text: 'New alert',
      type: 'info',
      url: 'url',
      urlCaption: 'caption'
    }
  },
  result: {
    data: {
      addAlert: [
        {
          id: '1',
          test: 'New alert',
          type: 'info',
          url: 'url',
          urlCaption: 'caption'
        }
      ]
    }
  }
}

describe('AdminLewAlert component', () => {
  test('Should create new alert', async () => {
    const { container } = render(
      <MockedProvider mocks={[addAlertMock]} addTypename={false}>
        <NewAlert setAlerts={() => {}} />
      </MockedProvider>
    )
    expect(container).toMatchSnapshot()
    await userEvent.type(screen.getByTestId('input0'), 'New alert', {
      delay: 1
    })
    await userEvent.type(screen.getByTestId('input2'), 'url', { delay: 1 })
    await userEvent.type(screen.getByTestId('input3'), 'caption', { delay: 1 })
    userEvent.click(screen.getByText('Create New Alert'))
    await waitFor(() => expect(screen.getByTestId('input0').value).toEqual(''))
    await waitFor(() => expect(screen.getByTestId('input2').value).toEqual(''))
    await waitFor(() => expect(screen.getByTestId('input3').value).toEqual(''))
  })
  test('Should capture error', async () => {
    const errorAlertMock = { ...addAlertMock, error: new Error('fail') }
    delete errorAlertMock.result
    render(
      <MockedProvider mocks={[errorAlertMock]} addTypename={false}>
        <NewAlert setAlerts={() => {}} />
      </MockedProvider>
    )
    await userEvent.type(screen.getByTestId('input0'), 'New alert', {
      delay: 1
    })
    await userEvent.type(screen.getByTestId('input2'), 'url', { delay: 1 })
    await userEvent.type(screen.getByTestId('input3'), 'caption', { delay: 1 })
    userEvent.click(screen.getByText('Create New Alert'))
    await waitFor(() => expect(Sentry.captureException).toBeCalled())
  })
  test('Should refuse creating new alert with incomplete data', async () => {
    const { container } = render(
      <MockedProvider mocks={[addAlertMock]} addTypename={false}>
        <NewAlert setAlerts={() => {}} />
      </MockedProvider>
    )
    await userEvent.type(screen.getByTestId('input2'), 'url', { delay: 1 })
    await userEvent.type(screen.getByTestId('input3'), 'caption', { delay: 1 })
    userEvent.click(screen.getByText('Create New Alert'))
    await waitFor(() => expect(screen.getByText('Required')).toBeTruthy())
    expect(container).toMatchSnapshot()
  })
})
