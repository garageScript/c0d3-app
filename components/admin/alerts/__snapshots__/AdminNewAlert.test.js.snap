// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminLewAlert component Should create new alert 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <div
      class="col-8 offset-2"
    >
      <div
        class="mb-3"
      >
        <div
          class="alert d-flex justify-content-between mt-3 bg-primary text-white"
          role="alert"
        >
          <div
            class="d-flex gap-3"
          >
            <img
              height="24"
              src="/assets/curriculum/icons/icon-tip.svg"
              width="24"
            />
             
            
          </div>
          <button
            class="alert-dismiss"
            data-testid="dismiss-info"
            role="dismiss"
          >
            <img
              src="/assets/curriculum/icons/dismiss-info.svg"
            />
          </button>
        </div>
      </div>
      <div
        class="text-start"
      >
        <div
          class="row"
        >
          <div
            class="card shadow-sm col-12"
          >
            <div
              class="card-body text-center"
            />
            <div
              class="text-start"
            >
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5text0"
                >
                  Text
                </h5>
                <input
                  class="form-control"
                  data-testid="input0"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5type1"
                >
                  Type
                </h5>
                <div
                  class="d-none d-lg-block"
                >
                  <div
                    class="dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle title title-none"
                      type="button"
                    >
                      info
                    </button>
                  </div>
                </div>
                <div
                  class="d-lg-none"
                >
                  <div
                    class="text-center px-4"
                  >
                    <button
                      class="dropdown-item"
                      data-rr-ui-dropdown-item=""
                    >
                      info
                    </button>
                  </div>
                  <div
                    class="text-center px-4"
                  >
                    <button
                      class="dropdown-item"
                      data-rr-ui-dropdown-item=""
                    >
                      urgent
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5url2"
                >
                  Url
                </h5>
                <input
                  class="form-control"
                  data-testid="input2"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5urlCaption3"
                >
                  Urlcaption
                </h5>
                <input
                  class="form-control"
                  data-testid="input3"
                  placeholder=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="text-center mb-4"
            >
              <button
                class="btn borderless btn-primary"
                style="color: rgb(255, 255, 255);"
              >
                Create New Alert
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdminLewAlert component Should refuse creating new alert with incomplete data 1`] = `
<div>
  <div
    class="row mb-4"
  >
    <div
      class="col-8 offset-2"
    >
      <div
        class="mb-3"
      >
        <div
          class="alert d-flex justify-content-between mt-3 bg-primary text-white"
          role="alert"
        >
          <div
            class="d-flex gap-3"
          >
            <img
              height="24"
              src="/assets/curriculum/icons/icon-tip.svg"
              width="24"
            />
             
            <a
              class="text-white noUnderline"
              href="url"
              rel="noopener noreferrer"
              target="_blank"
            >
              caption
            </a>
          </div>
          <button
            class="alert-dismiss"
            data-testid="dismiss-info"
            role="dismiss"
          >
            <img
              src="/assets/curriculum/icons/dismiss-info.svg"
            />
          </button>
        </div>
      </div>
      <div
        class="text-start"
      >
        <div
          class="row"
        >
          <div
            class="card shadow-sm col-12"
          >
            <div
              class="card-body text-center"
            />
            <div
              class="text-start"
            >
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5text0"
                >
                  Text
                </h5>
                <input
                  class="form-control"
                  data-testid="input0"
                  placeholder=""
                  type="text"
                  value=""
                />
                <h6
                  class="text-danger mb-0"
                >
                  Required
                </h6>
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5type1"
                >
                  Type
                </h5>
                <div
                  class="d-none d-lg-block"
                >
                  <div
                    class="dropdown"
                  >
                    <button
                      aria-expanded="false"
                      class="dropdown-toggle title title-none"
                      type="button"
                    >
                      info
                    </button>
                  </div>
                </div>
                <div
                  class="d-lg-none"
                >
                  <div
                    class="text-center px-4"
                  >
                    <button
                      class="dropdown-item"
                      data-rr-ui-dropdown-item=""
                    >
                      info
                    </button>
                  </div>
                  <div
                    class="text-center px-4"
                  >
                    <button
                      class="dropdown-item"
                      data-rr-ui-dropdown-item=""
                    >
                      urgent
                    </button>
                  </div>
                </div>
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5url2"
                >
                  Url
                </h5>
                <input
                  class="form-control"
                  data-testid="input2"
                  placeholder=""
                  type="text"
                  value="url"
                />
              </div>
              <div
                class="d-flex flex-column ms-3 me-3 mb-4"
              >
                <h5
                  data-testid="h5urlCaption3"
                >
                  Urlcaption
                </h5>
                <input
                  class="form-control"
                  data-testid="input3"
                  placeholder=""
                  type="text"
                  value="caption"
                />
              </div>
            </div>
            <div
              class="text-center mb-4"
            >
              <button
                class="btn borderless btn-primary"
                style="color: rgb(255, 255, 255);"
              >
                Create New Alert
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
